### PR TITLE
Add kwargs to optimizer, scheduler and dataloader using function `accelerator().load_state()`

### PR DESCRIPTION
**Pull request by luiz0992**
# What does this PR do?

 This PR fixes issue #3539
  
This PR adds support for passing keyword arguments to the load functions of optimizers, schedulers, and dataloaders when using the Accelerator's load_state method. This enhancement allows for more
  flexibility and control when loading optimizer, scheduler, and dataloader states.

  The changes include:
  1. Extending the load_state method signature to accept dedicated kwargs dictionaries for each component
  2. Passing these kwargs to the respective components' load functions
  3. Updating docstrings to document the new parameters
  

## Before submitting
- [x] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/accelerate/blob/main/CONTRIBUTING.md#submitting-a-pull-request-pr),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/accelerate/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/accelerate/tree/main/docs#writing-documentation---specification).
- [x] Did you write any new necessary tests?

## Who can review?

 @BenjaminBossan @SunMarc @zach-huggingface

### CODE REVIEW

**Review by SunMarc — COMMENTED** [review]
Thanks ! Shared about another solution in the issue.

File: src/accelerate/checkpointing.py [hunk 1]
```diff
@@ -200,6 +203,12 @@ def load_accelerator_state(
             An optional *GradScaler* instance to load
         map_location (`str`, *optional*):
             What device to load the optimizer state onto. Should be one of either "cpu" or "on_device".
+        optimizer_load_kwargs (`dict`, *optional*):
+            Additional arguments that can be passed to the optimizer's `load` function.
+        scheduler_load_kwargs (`dict`, *optional*):
+            Additional arguments that can be passed to the scheduler's `load` function.
+        dataloader_load_kwargs (`dict`, *optional*):
+            Additional arguments that can be passed to the dataloader's `load` function.
```
Comment (SunMarc) [inline src/accelerate/checkpointing.py:211]: instead of create kwargs for each one of them, let's just have `load_kwargs` that we use for all load method in this function. Also it is better to set `load_kwargs` to None by default and in the function set it to dict() as it is mutable.

File: src/accelerate/checkpointing.py [hunk 2]
```diff
@@ -261,7 +270,7 @@ def load_accelerator_state(
             dataloader_state_dict_name = "dl_state_dict.bin" if i == 0 else f"dl_state_dict_{i}.bin"
             input_dataloader_state_dict_file = input_dir.joinpath(dataloader_state_dict_name)
             if input_dataloader_state_dict_file.exists():
-                state_dict = load(input_dataloader_state_dict_file)
+                state_dict = load(input_dataloader_state_dict_file, map_location=None, **dataloader_load_kwargs)
                 dataloader.load_state_dict(state_dict)
     logger.info("All dataloader sampler states loaded successfully")
```
Comment (SunMarc) [inline src/accelerate/checkpointing.py:276]: let's also include it for scaler and states

**Review by SunMarc — COMMENTED** [review]
Thanks ! just some nits

**Comment by SunMarc** [comment]
@luiz0992, let me know if you are planning to finish this pr or not

File: src/accelerate/checkpointing.py (hunk 1 above)
Comment (luiz0992) [reply src/accelerate/checkpointing.py]: I am not sure if I understood correct your suggestions. You want to use `load_kwargs`  for all load functions? I implemented that but I am not sure if this is the best way to go. On my case the error only happend with the optimizer load. Anyway, let me know what you think

810c69eed6a5759cecf80b70f31ceb39c15b854b

File: src/accelerate/checkpointing.py (hunk 2 above)
Comment (luiz0992) [reply src/accelerate/checkpointing.py:267]: 810c69eed6a5759cecf80b70f31ceb39c15b854b

File: src/accelerate/accelerator.py [hunk 3]
```diff
@@ -3357,10 +3357,7 @@ def register_load_state_pre_hook(self, hook: Callable[..., None]) -> hooks.Remov
     def load_state(
         self,
         input_dir: str = None,
-        optimizer_load_kwargs: dict[str, Any] = {},
-        scheduler_load_kwargs: dict[str, Any] = {},
-        dataloader_load_kwargs: dict[str, Any] = {},
-        **load_model_func_kwargs,
```
[the remark this answers is not in the document]
Comment (luiz0992) [reply src/accelerate/accelerator.py]: The `load_model` does not accept kwargs. Let me know if I am mistaken.

[the remark this answers is not in the document]
Comment (SunMarc) [reply src/accelerate/accelerator.py]: in safetensors, we have the following
```python
def load_model(
    model: torch.nn.Module, filename: Union[str, os.PathLike], strict: bool = True, device: Union[str, int] = "cpu"
) -> Tuple[List[str], List[str]]:
```
If you can revert the changes related to `load_model_func_kwargs` and only update `load_kwargs` to where we use `load`, it will be better

[the remark this answers is not in the document]
Comment (luiz0992) [reply src/accelerate/accelerator.py]: The way it is being used it is missing only the `strict` argument, but that's ok. I reverted the `load_model_func_kwargs`.

c3d7f1fef1430c4a786dafd38941871175ac7e6c

**Review by SunMarc — APPROVED** [review]
Thanks !

File: src/accelerate/checkpointing.py [hunk 4]
```diff
@@ -180,6 +180,7 @@ def load_accelerator_state(
     process_index,
     scaler=None,
     map_location=None,
+    load_kwargs=None,
```
Comment (SunMarc) [inline src/accelerate/checkpointing.py:183]: just make sure to init a dict afterwards, toherwise, it will raise an error: 

```python 
if load_kwargs is None:
        load_kwargs = {}
```

**Review by SunMarc — COMMENTED** [review]
Just a small nit

File: src/accelerate/checkpointing.py (hunk 4 above)
Comment (luiz0992) [reply src/accelerate/checkpointing.py:183]: e54694ce06cc74716a375f33f72436e40a4a9a64
